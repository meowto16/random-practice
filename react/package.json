{
  "name": "react-experiments",
  "version": "0.1.0",
  "private": true,
  "dependencies": {
    "@hookform/resolvers": "^2.9.8",
    "@vkontakte/icons": "^1.192.0",
    "@vkontakte/vk-bridge": "^2.7.0",
    "@vkontakte/vkui": "^4.37.2",
    "axios": "^0.27.2",
    "nock": "^13.2.9",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "react-hook-form": "^7.35.0",
    "react-scripts": "5.0.1",
    "web-vitals": "^2.1.4",
    "yup": "^0.32.11"
  },
  "scripts": {
    "start": "react-scripts start",
    "build": "react-scripts build",
    "test": "react-scripts test",
    "eject": "react-scripts eject"
  },
  "eslintConfig": {
    "extends": [
      "react-app",
      "react-app/jest"
    ]
  },
  "browserslist": {
    "production": [
      ">0.2%",
      "not dead",
      "not op_mini all"
    ],
    "development": [
      "last 1 chrome version",
      "last 1 firefox version",
      "last 1 safari version"
    ]
  },
  "devDependencies": {
    "@faker-js/faker": "^7.5.0",
    "@testing-library/jest-dom": "^5.16.5",
    "@testing-library/react": "^13.4.0",
    "@testing-library/user-event": "^13.5.0"
  }
}
