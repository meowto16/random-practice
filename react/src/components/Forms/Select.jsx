import { FormItem, Select as VKSelect } from '@vkontakte/vkui'
import { useFormContext } from 'react-hook-form'

export const Select = ({ FormItemProps, SelectProps, name, label, placeholder, options }) => {
  const { register, formState, getFieldState } = useFormContext()

  const { onBlur, onChange, name: registerName, ref }  = register(name);

  const field = getFieldState(name, formState)

  return (
    <FormItem
      {...FormItemProps}
      top={label || FormItemProps?.top}
      bottom={field.error ? field.error.message : FormItemProps?.bottom}
      status={Boolean(field.error) ? 'error' : 'default'}
    >
      <VKSelect
        {...SelectProps}
        aria-label={label}
        name={registerName}
        placeholder={placeholder}
        onBlur={onBlur}
        onChange={onChange}
        getRef={ref}
        options={options}
      />
    </FormItem>
  )
}