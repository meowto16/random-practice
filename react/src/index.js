import { ConfigProvider } from '@vkontakte/vkui'

import '@vkontakte/vkui/dist/vkui.css'
import React from 'react'
import ReactDOM from 'react-dom/client'

import App from './App'
import './index.css'

const root = ReactDOM.createRoot(document.getElementById('root'));
root.render(
  <React.StrictMode>
    <ConfigProvider appearance="dark" isWebView={false} platform="ios">
      <App />
    </ConfigProvider>
  </React.StrictMode>
);
