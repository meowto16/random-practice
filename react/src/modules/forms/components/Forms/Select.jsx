import { Select as VKSelect } from '@vkontakte/vkui'
import { useCallback } from 'react'
import { useFormContext } from 'react-hook-form'

import { FormItem } from './FormItem'

export const Select = ({ FormItemProps, SelectProps, name, label, placeholder, options }) => {
  const form = useFormContext()
  const value = form.watch(name)

  const handleChange = useCallback((event) => {
    form.setValue(name, event.target.value ?? undefined)
  }, [form, name])

  return (
    (
      <FormItem
        {...FormItemProps}
        name={name}
        label={label}
      >
        {({ name, label, onBlur, getRef: ref }) => (
          <VKSelect
            {...SelectProps}
            name={name}
            label={label}
            aria-label={label}
            placeholder={placeholder}
            options={options}
            value={value}
            onChange={handleChange}
            onBlur={onBlur}
            getRef={ref}
          />
        )}
      </FormItem>
    )
  )
}