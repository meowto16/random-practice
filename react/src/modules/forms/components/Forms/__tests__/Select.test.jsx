import { act, findByText, fireEvent } from '@testing-library/react'
import * as yup from 'yup'
import selectEvent from 'react-select-event'

import { Select } from '../Select'

import { renderWithReactHookForm } from '../../../../utils.testing'

const eventMock = expect.anything();

describe('Select (Form component)', () => {
  const options = [
    { value: 'russia', label: 'Россия' },
    { value: 'ukraine', label: 'Украина' },
    { value: 'germany', label: 'Германия' },
    { value: 'france', label: 'Франция' },
  ]


  it('renders without errors', () => {
    const { screen } = renderWithReactHookForm(<Select name="country" label="Страна" options={options} />);
    screen.unmount();
  })

  it('same as snapshot', () => {
    const { screen } = renderWithReactHookForm(<Select name="country" label="Страна" options={options} />);
    expect(screen.container.firstChild).toMatchSnapshot();
  })

  it.skip('supports default value', async () => {
    const { screen } = renderWithReactHookForm(
      <Select name="country" label="Страна" />,
      {
        defaultValues: {
          country: 'russia'
        }
      });

    expect( await screen.findByText('Россия')).toBeInTheDocument()
  })

  it.skip('changes form value (react-hook-form context)', async () => {
    const { screen, form, submitHandler } = renderWithReactHookForm(<Select name="country" label="Страна" placeholder="Выберите страну" />)

    // Check empty value
    await act(() => {
      fireEvent.submit(form)
    })
    expect(submitHandler).toHaveBeenCalledWith({ country: undefined }, eventMock)

    // Change value
    const select = screen.getByRole('select')
    await act(() => {
      selectEvent.select(select, 'Франция')
      fireEvent.submit(form)
    })

    // Check changed value
    expect(submitHandler).toHaveBeenCalledWith({ country: 'Франция' }, eventMock)
  })

  it.skip('supports validation', async () => {
    const { screen, form } = renderWithReactHookForm(
      <Select name="country" label="Страна" placeholder="Выберите страну" />,
      {
        validationSchema: yup.object({
          firstname: yup.string().required('Это обязательное поле')
        }),
      }
    )

    // Error is not visible before submit form
    expect(screen.queryByText('Это обязательное поле')).not.toBeTruthy()

    // Submit form
    await act(() => void fireEvent.submit(form))

    // Error is now visible after submit form
    expect(screen.queryByText('Это обязательное поле')).toBeTruthy()
  })
})