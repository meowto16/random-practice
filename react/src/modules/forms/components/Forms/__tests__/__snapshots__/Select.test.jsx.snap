// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select (Form component) same as snapshot 1`] = `
<form
  name="form"
>
  <div
    class="vkuiFormItem vkuiFormItem--android FormItem FormItem--android vkuiFormItem--default FormItem--default vkuiFormItem--sizeY-regular FormItem--sizeY-regular vkuiFormItem--withTop FormItem--withTop"
  >
    <h5
      class="vkuiFormItem__top FormItem__top vkuiSubhead Subhead vkuiSubhead--sizeY-regular Subhead--sizeY-regular"
    >
      Страна
    </h5>
    <label
      class="vkuiCustomSelect CustomSelect"
    >
      <div
        aria-hidden="true"
        aria-label="Страна"
        class="vkuiSelect vkuiSelect--android Select Select--android vkuiSelect--default Select--default vkuiSelect--empty Select--empty vkuiSelect--sizeX-compact Select--sizeX-compact vkuiSelect--sizeY-regular Select--sizeY-regular vkuiSelect--hasAfter Select--hasAfter vkuiFormField FormField vkuiFormField--default FormField--default vkuiFormField--status-default FormField--status-default vkuiFormField--sizeY-regular FormField--sizeY-regular"
        label="Страна"
        role="presentation"
        tabindex="0"
      >
        <div
          class="vkuiSelect__container Select__container"
        >
          <span
            class="vkuiSelect__title Select__title vkuiHeadline vkuiHeadline--android Headline Headline--android vkuiHeadline--sizeY-regular Headline--sizeY-regular vkuiHeadline--l-1 Headline--l-1 vkuiHeadline--w-3 Headline--w-3"
          />
        </div>
        <div
          class="vkuiFormField__after FormField__after"
          role="presentation"
        >
          <div
            class=" Icon Icon--24 Icon--w-24 Icon--h-24 Icon--chevron_down_24 vkuiDropdownIcon DropdownIcon"
            role="presentation"
            style="width: 24px; height: 24px;"
          >
            <svg
              height="24"
              style="display: block;"
              viewBox="0 0 24 24"
              width="24"
            >
              <use
                style="fill: currentColor;"
                xlink:href="#chevron_down_24"
              />
            </svg>
          </div>
        </div>
        <div
          class="vkuiFormField__border FormField__border"
          role="presentation"
        />
      </div>
      <select
        aria-hidden="true"
        class="vkuiCustomSelect__control CustomSelect__control"
        name="country"
      >
        <option
          value="russia"
        />
        <option
          value="ukraine"
        />
        <option
          value="germany"
        />
        <option
          value="france"
        />
      </select>
    </label>
  </div>
  <button
    type="submit"
  >
    Отправить
  </button>
</form>
`;
